File: implementations/java-springboot/harness/src/rules/SharedInfra.java
package harness.rules;

import harness.Finding;
import harness.RuleResult;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import static harness.JavaFiles.collectJavaFiles;
import static harness.JavaFiles.readText;

/**
 * [7] shared-infra: outbox / task-queue
 *
 * The outbox trigger is decided by "is there code that actually references OutboxWriter" —
 * this avoids false-triggering on an unrelated file whose name happens to contain
 * "Outbox" (a past bug: the real files were already all inside outbox/, so the "find a
 * file outside it" condition was always false, meaning it only ever emitted SKIP and never
 * actually validated the outbox package). OutboxWriter is the "loads events into the
 * outbox table" component that exists regardless of whether synchronous draining (the old
 * approach) or Poller/Consumer-based async draining (current) is used, so it's used as the
 * trigger — OutboxRelay was removed in the switch to async, so it can no longer be used as
 * the trigger.
 */
public final class SharedInfra {
    private SharedInfra() {
    }

    // build/ mirrors the package structure of the compiled .class files, so without
    // excluding it, findDirsNamed would find the same-named directory twice — once under
    // src and once under build (excluded for the same reason as in JavaFiles.java).
    private static final Set<String> EXCLUDED_DIRS = Set.of("test", ".git", "build");

    public static RuleResult check(String rootPath) {
        File root = new File(rootPath);
        RuleResult result = new RuleResult("shared-infra");
        for (Finding f : checkOutboxPattern(root)) result.add(f);
        for (Finding f : checkTaskQueuePattern(root)) result.add(f);
        return result;
    }

    private static List<Finding> checkOutboxPattern(File root) {
        boolean usesOutboxWriter = collectJavaFiles(root).stream()
            .anyMatch(f -> readText(f).contains("OutboxWriter"));
        if (!usesOutboxWriter) {
            return List.of(Finding.skip("No outbox pattern"));
        }

        List<File> outboxDirs = new ArrayList<>();
        findDirsNamed(root, "outbox", outboxDirs);

        if (outboxDirs.isEmpty()) {
            return List.of(Finding.fail("outbox package", "References OutboxWriter but has no outbox/ package"));
        }

        boolean hasWriter = outboxDirs.stream().anyMatch(d -> new File(d, "OutboxWriter.java").isFile());
        boolean hasPoller = outboxDirs.stream().anyMatch(d -> new File(d, "OutboxPoller.java").isFile());
        boolean hasConsumer = outboxDirs.stream().anyMatch(d -> new File(d, "OutboxConsumer.java").isFile());
        if (hasWriter && hasPoller && hasConsumer) {
            return List.of(Finding.pass("outbox package (confirmed OutboxWriter/OutboxPoller/OutboxConsumer implementations)"));
        }
        return List.of(Finding.fail("outbox package", "outbox/ package exists but one or more of OutboxWriter.java/OutboxPoller.java/OutboxConsumer.java could not be found — outbox load (Writer) + queue publish (Poller) + queue consume (Consumer) must all be present (domain-events.md)"));
    }

    private static List<Finding> checkTaskQueuePattern(File root) {
        boolean hasTaskFile = collectJavaFiles(root).stream()
            .anyMatch(f -> f.getName().contains("TaskQueue"));
        if (!hasTaskFile) {
            return List.of(Finding.skip("No task-queue pattern"));
        }

        List<File> taskDirs = new ArrayList<>();
        findDirsNamed(root, "task-queue", taskDirs);
        findDirsNamed(root, "taskqueue", taskDirs);

        if (!taskDirs.isEmpty()) {
            return List.of(Finding.pass("task-queue package"));
        }
        return List.of(Finding.fail("task-queue package", "TaskQueue file exists but no task-queue/ package"));
    }

    private static void findDirsNamed(File dir, String name, List<File> out) {
        if (!dir.exists()) return;
        if (dir.isDirectory() && dir.getName().equals(name)) {
            out.add(dir);
        }
        File[] children = dir.listFiles();
        if (children == null) return;
        for (File child : children) {
            if (child.isDirectory() && !EXCLUDED_DIRS.contains(child.getName())) {
                findDirsNamed(child, name, out);
            }
        }
    }
}

File: implementations/kotlin-springboot/harness/src/rules/SharedInfra.kt
package harness.rules

import harness.*
import java.io.File

/**
 * [8] shared-infra: outbox / task-queue
 *
 * The outbox trigger is determined by "is there code that actually references OutboxWriter" — to
 * avoid being tripped up by an unrelated file whose name happens to contain "Outbox"(a past bug: the
 * actual files were already entirely inside outbox/, so the "find a file outside it" condition was
 * always false, meaning it only ever SKIPped and never substantively verified the outbox package).
 *
 * `OutboxRelay`(synchronous drain) is not present — Outbox → queue publishing is handled by
 * `OutboxPoller` and queue → handler routing by `OutboxConsumer`. This rule's trigger and
 * required-file checks target that current design(requiring OutboxRelay.kt would always fail now).
 */
fun checkSharedInfra(rootPath: String): RuleResult {
    val root = File(rootPath)
    val result = RuleResult("shared-infra")
    checkOutboxPattern(root).forEach { result.add(it) }
    checkTaskQueuePattern(root).forEach { result.add(it) }
    return result
}

private fun checkOutboxPattern(root: File): List<Finding> {
    val usesOutboxWriter = collectKtFiles(root).any { it.readText().contains("OutboxWriter") }
    if (!usesOutboxWriter) return listOf(skipFinding("no outbox pattern"))

    val outboxDirs = root.walkTopDown()
        .onEnter { it.name !in setOf("test", "build") }
        .filter { it.isDirectory && it.name == "outbox" }
        .toList()

    if (outboxDirs.isEmpty()) {
        return listOf(failFinding("outbox package", "references OutboxWriter but there is no outbox/ package"))
    }

    val hasWriter = outboxDirs.any { File(it, "OutboxWriter.kt").isFile }
    val hasPoller = outboxDirs.any { File(it, "OutboxPoller.kt").isFile }
    val hasConsumer = outboxDirs.any { File(it, "OutboxConsumer.kt").isFile }
    return if (hasWriter && hasPoller && hasConsumer) {
        listOf(passFinding("outbox package (confirmed OutboxWriter/OutboxPoller/OutboxConsumer implementations)"))
    } else {
        listOf(
            failFinding(
                "outbox package",
                "the outbox/ package exists but OutboxWriter.kt/OutboxPoller.kt/OutboxConsumer.kt could not be found",
            ),
        )
    }
}

private fun checkTaskQueuePattern(root: File): List<Finding> {
    val hasTaskFile = root.walkTopDown()
        .onEnter { it.name !in setOf("test", "build") }
        .any { it.isFile && it.extension == "kt" && it.name.contains("TaskQueue") }

    if (!hasTaskFile) return listOf(skipFinding("no task-queue pattern"))

    val hasTaskDir = root.walkTopDown()
        .onEnter { it.name !in setOf("test", "build") }
        .any { it.isDirectory && (it.name == "task-queue" || it.name == "taskqueue") }

    return if (hasTaskDir) {
        listOf(passFinding("task-queue package"))
    } else {
        listOf(failFinding("task-queue package", "a TaskQueue file exists but there is no task-queue/ package"))
    }
}
